File: mobile/library/swift/src/RequestMapper.swift
private let kRestrictedHeaderPrefixes = [":", "x-envoy-mobile"]

extension Request {
  /// Returns a set of outbound headers that include HTTP
  /// information on the URL, method, and additional headers.
  ///
  /// - returns: Outbound headers to send with an HTTP request.
  func outboundHeaders() -> [String: [String]] {
    var headers = self.headers
      .filter { !kRestrictedHeaderPrefixes.contains(where: $0.key.hasPrefix) }
      .reduce(into: [
        ":method": [self.method.stringValue],
        ":scheme": [self.scheme],
        ":authority": [self.authority],
        ":path": [self.path],
      ]) { $0[$1.key] = $1.value }

    if let retryPolicy = self.retryPolicy {
      headers = headers.merging(retryPolicy.outboundHeaders()) { _, retryHeader in retryHeader }
    }

    if let upstreamHttpProtocol = self.upstreamHttpProtocol {
      headers["x-envoy-mobile-upstream-protocol"] = [upstreamHttpProtocol.stringValue]
    }

    return headers
  }
}

File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/RequestMapper.kt
package io.envoyproxy.envoymobile

internal fun Request.outboundHeaders(): Map<String, List<String>> {
  val retryPolicyHeaders = retryPolicy?.outboundHeaders() ?: emptyMap()

  val result = mutableMapOf<String, List<String>>()
  result.putAll(
    headers.filter { entry ->
      !entry.key.startsWith(":") && !entry.key.startsWith("x-envoy-mobile")
    }
  )
  result.putAll(retryPolicyHeaders)
  result[":method"] = listOf(method.stringValue)
  result[":scheme"] = listOf(scheme)
  result[":authority"] = listOf(authority)
  result[":path"] = listOf(path)

  if (upstreamHttpProtocol != null) {
    result["x-envoy-mobile-upstream-protocol"] = listOf(upstreamHttpProtocol.stringValue)
  }

  return result
}
